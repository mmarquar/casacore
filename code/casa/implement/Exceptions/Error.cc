//# Error.cc: Base class for all AIPS++ errors (templated classes)
//# Copyright (C) 1993,1994,1995,1996
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/Exceptions/Error.h>
#include <iostream.h>


//#  This template is for index errors  where the Index can be any type. It
//#  is derived from the class IndexError. This derivation allows the user
//#  to catch all index errors with one catch statement. The remainder of
//#  of the class is very similar to the AipsError class.
rtti_imp_init_a1(indexError);
rtti_imp_mbrf_a1(indexError);

template<class t>
indexError<t>::indexError(t oI, const Char *str) : oIndex(oI), IndexError(str)
{}
template<class t>
indexError<t>::indexError(t oI, const String &str) : oIndex(oI), IndexError(str)
{}

template<class t>
indexError<t>::indexError(ExcpError *excp) : IndexError(excp)
{
    indexError<t> *tmp;
    PCAST(tmp,indexError<t>,excp);
    if (tmp) {
        oIndex = (*tmp).oIndex;
        _equal = True;
    } else {
        _equal = False;
    }
}

template<class t>
indexError<t>::~indexError()
    {}

//#  This template is for duplicate key errors  where the Key can be any type. It
//#  is derived from the class DuplError. This derivation allows the user
//#  to catch all index errors with one catch statement. The remainder of
//#  of the class is very similar to the AipsError class.
rtti_imp_init_a1(duplError);
rtti_imp_mbrf_a1(duplError);

template<class t>
duplError<t>::duplError(ExcpError *excp) : DuplError(excp)
{
    duplError<t> *tmp;
    PCAST(tmp,duplError<t>,excp);
    if (tmp) {
        oKey = (*tmp).oKey;
        _equal = True;
    } else {
        _equal = False;
    }
}

template<class t>
duplError<t>::~duplError()
    {}


template<class t>
duplError<t>::duplError(t oI, const Char *str) : oKey(oI), DuplError(str)
{}
template<class t>
duplError<t>::duplError(t oI, const String &str) : oKey(oI), DuplError(str)
{}

