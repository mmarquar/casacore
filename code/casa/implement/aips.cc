//# aips.cc: Global initialization for namespace management, standard types, etc.
//# Copyright (C) 1993,1994,1996,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/aips.h>
#include <stdio.h>
#include <stdlib.h>

#if defined(AIPS_DEBUG)
Bool aips_debug_on = True;
#else
Bool aips_debug_on = False;
#endif

static size_t total = 0;
const size_t &aipsalloc(total);

// Replace global operator new/delete unless the user says to do otherwise.
// We do this to keep track of memory use.
#if !defined(AIPS_STDNEW)
#include <new.h>

// Assume everything aligns on 8 byte boundaries. No portable way to determine
// this. We store the allocation size "before" the new'd pointer in an area this
// long
const int offset = 8;

// When we have modern compilers we should throw an exception if the new_handler
// fails.

void *operator new (size_t size)
{
    char *ptr;
    while ((ptr = (char *)malloc(size+offset)) == 0) {
	new_handler handler;
	handler = set_new_handler(0);
	set_new_handler(handler); // Put it back before trying it
	if (handler) {
	    (*handler)();
	} else {
	    return 0;
	}
    }
    *((size_t *)ptr) = size;
    if (sizeof(size_t)*2 <= offset) {
	// Write it twice if it will fit as a sanity check.
	*((size_t *)ptr+1) = size;
    }
    total += size;
    return ptr+offset;
}

void operator delete (void *ptr)
{
    if (ptr) {
	char *alias = (char *)ptr;
	alias -= offset;
	size_t size = *((size_t*)alias);
	if (sizeof(size_t)*2 <= offset) {
	    if (size != *((size_t*)alias+1)) {
		fprintf(stderr, "ERROR: delete'ing memory not allocated "
			"with new!\n");
		exit(999);
	    }
	}
	total -= size;
	free(alias);
    }
}

//------ Just copies of the above with [] added

void *operator new[] (size_t size)
{
    char *ptr;
    while ((ptr = (char *)malloc(size+offset)) == 0) {
	new_handler handler;
	handler = set_new_handler(0);
	set_new_handler(handler); // Put it back before trying it
	if (handler) {
	    (*handler)();
	} else {
	    return 0;
	}
    }
    *((size_t *)ptr) = size;
    if (sizeof(size_t)*2 <= offset) {
	// Write it twice if it will fit as a sanity check.
	*((size_t *)ptr+1) = size;
    }
    total += size;
    return ptr+offset;
}

void operator delete[] (void *ptr)
{
    if (ptr) {
	char *alias = (char *)ptr;
	alias -= offset;
	size_t size = *((size_t*)alias);
	if (sizeof(size_t)*2 <= offset) {
	    if (size != *((size_t*)alias+1)) {
		fprintf(stderr, "ERROR: delete'ing memory not allocated "
			"with new!\n");
		exit(999);
	    }
	}
	total -= size;
	free(alias);
    }
}

#endif
