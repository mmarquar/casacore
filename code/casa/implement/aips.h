//# aips.h: Global initialization for namespace management, standard types, etc.
//# Copyright (C) 1993-1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#if !defined(AIPS_ENTER_)
#include <aips/aips_enter.h>
#endif

#if !defined(AIPS_AIPS_H)
#define AIPS_AIPS_H

#if defined(_AIX)
#pragma implementation ("aips.cc")
#endif

// At least the centerline compiler doesn't know what to do with volatile.
// This may be more generally true.

#if defined(OBJECTCENTER)
#define volatile
                    // nothing
#endif

// Including aips.h should guarantee that AIPS++ namespace management is
// switched on - it may have been explicitly switched off beforehand by
// aips_exit.h.

// Defined the "aips_" macro which is used for name space management, and
// the "aips_name2" macro which is used to join two tokens.

#if defined(__STDC__) || defined(__ANSI_CPP__) || defined(__hpux)
#define aips_(x)  aips##x
#define aips_name2(a,b) a##b
#else
#define aips_(x)  aips/**/x
#define aips_name2(a,b) a/**/b
#endif

// This section contains the various standard types used by AIPS++.

enum Bool { False = 0, True = 1 };

typedef char Char;
typedef unsigned char uChar;
typedef short Short;
typedef unsigned short uShort;
typedef int Int;
typedef unsigned int uInt;
typedef long Long;
typedef unsigned long uLong;
typedef float Float;
typedef double Double;
#if defined (AIPS_IRIX) && !defined(__GNUG__)
typedef double lDouble;
#else
typedef long double lDouble;
#endif

#if defined(__GNUG__)
// Needed for g++ 2.7.2. Hopefully we can delete these at some point.
// <group>
inline const Bool &at_cc(const Bool &val) { return(val); };
inline const Char &at_cc(const Char &val) { return(val); };
inline const uChar &at_cc(const uChar &val) { return(val); };
inline const Short &at_cc(const Short &val) { return(val); };
inline const uShort &at_cc(const uShort &val) { return(val); };
inline const Int &at_cc(const Int &val) { return(val); };
inline const uInt &at_cc(const uInt &val) { return(val); };
inline const Long &at_cc(const Long &val) { return(val); };
inline const uLong &at_cc(const uLong &val) { return(val); };
inline const Float &at_cc(const Float &val) { return(val); };
inline const Double &at_cc(const Double &val) { return(val); };
inline const lDouble &at_cc(const lDouble &val) { return(val); };

inline Bool &at_c(Bool &val) { return(val); };
inline Char &at_c(Char &val) { return(val); };
inline uChar &at_c(uChar &val) { return(val); };
inline Short &at_c(Short &val) { return(val); };
inline uShort &at_c(uShort &val) { return(val); };
inline Int &at_c(Int &val) { return(val); };
inline uInt &at_c(uInt &val) { return(val); };
inline Long &at_c(Long &val) { return(val); };
inline uLong &at_c(uLong &val) { return(val); };
inline Float &at_c(Float &val) { return(val); };
inline Double &at_c(Double &val) { return(val); };
inline lDouble &at_c(lDouble &val) { return(val); };
// </group>
#else    /* !__GNUG__ */
#define at_c(X) X
#define at_cc(X) X
#endif   /* __GNUG__ */

// This is an inline guaranteed-correct conversion function which converts
// to Bool any type which has meaning in a C++ logical expression.
template <class T>
inline
Bool ToBool (T val)
{
    return (val ? True : False);
}

// If AIPS_DEBUG is not defined, then the symbol expands to (0) which in an
// if should be removed by the dead code eliminator of any optimizer; thus
// using this in your code should have no performance penalty in the normal
// case. If compiled with AIPS_DEBUG, then aips_debug is (defined to )
// a global boolean variable (so it can be turned on and off in a debugger) 
// which is initialized to True.

extern Bool aips_debug_on;

#if !defined(AIPS_DEBUG)
#define aips_debug (0)
#else
// The reason that we just don't make this a variable here is so that
// we can link against libraries compiled with or without AIPS_DEBUG
// without having any missing symbols.
#define aips_debug aips_debug_on
#endif

// This is just syntactic sugar to tell the namespace management that a
// particular class is not under control of namespace management. The only
// place this will normally be needed is forward declarations of imported
// classes, e.g. from InterViews.
#define imported /* Nothing */

// These are new keywords in the upcoming C++ standard.
// They are defined to an empty string, so they can be used until
// all compilers interpret them.
// Especially the keyword explicit may give useful information.
// It can be given at a one-argument constructor telling the compiler
// that it should not be used for automatic conversions.
// E.g.    explicit IPosition (uInt);
#define explicit /* Nothing */
#define typename /* Nothing */

#endif
