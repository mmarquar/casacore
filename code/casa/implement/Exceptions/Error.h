//# Error.h: Base class for all AIPS++ errors
//# Copyright (C) 1993,1994,1995,1999,2000
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#if !defined(AIPS_ERROR_H)
#define AIPS_ERROR_H


#include <aips/aips_exit.h>
#include <sys/types.h>
#include <aips/aips.h>
#include <aips/Exceptions/Excp.h>
#include <aips/Utilities/String.h>

// <summary>Base class for all AIPS++ library errors</summary>
// <use visibility=export>
//
// <reviewed reviewer="" date="" tests="" demos="">
// </reviewed>
//
// <prerequisite>
//   <li> ExcpError
// </prerequisite>
//
// <synopsis>
//  This is the base class for all of the AIPS++ error classes. Because
//  all of the errors have a common base class, any error can be caught
//  with a single catch statement.
//
//  This class has a string which allows error messages to be propagated.
//
//  <note role=tip> The string member must be handled very carefully because
//        string is also derived from cleanup, thus the 
//        <src>message.makePermanent()</src> call in the implementation of
//        the constructors. This prevents the String from being cleaned up
//        in the middle of an exception.
//
// </synopsis>
//
// <example>
// <srcblock>
//      throw(AipsError("SOME STRING"));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>
//
class AipsError : public ExcpError {
protected:
  String message;
public:
  //
  // Simply returns the stored error message.
  //
  const String &getMesg() const {return(message);}
  //
  // Creates an AipsError and initializes the error message from
  // the parameter
  // <group>
  AipsError(const Char *str=0);
  AipsError(const String &str);
  // </group>

  //
  // This constructor is used by the exception mecanism to perform
  // type checking.
  //
  AipsError(ExcpError *excp);

  //
  // Destructor which does nothing.
  //
  ~AipsError();
};

// <summary>Allocation errors</summary>
// <use visibility=export>
//
// <reviewed reviewer="" date="" tests="" demos="">
// </reviewed>
//
// <synopsis>
//
// This class is used for allocation errors. It adds an extra
// data item, the failed allocation size. Otherwise much the
// same as <src>AipsError</src>.
//
// </synopsis>
//
// <example>
// <srcblock>
//     throw(AllocError("ANY STRING",1024));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>
//
class AllocError : public AipsError {
protected:
  size_t Size;
public:
  //
  // This constructor takes the error message and the failed
  // allocation size.
  //
  // <group>
  AllocError(const Char *str, uInt sze) : AipsError(str), Size(sze) {}
  AllocError(const String &str, uInt sze) : AipsError(str), Size(sze)  {}
  // </group>

  //
  // This function returns the failed allocation size.
  //
  size_t size() {return(Size);}

  //
  // This constructor is used by the exception mecanism to perform
  // type checking.
  //
  AllocError(ExcpError *excp);
  
  //
  // Destructor which does nothing.
  //
  ~AllocError();

};

// <summary>Base class for all indexing errors</summary>
// <use visibility=export>
//
// <reviewed reviewer="" date="" tests="" demos="">
// </reviewed>
//
// <synopsis>
// This class is the base class of all <src>IndexError</src>s. It is
// defined to allow the user to catch any of the many kinds of IndexErrors
// which may be thrown. It can also be thrown itself if returning
// the illegal index value is unimportant.
// </synopsis>
//
// <example>
// <srcblock>
//     throw(IndexError("ANY STRING"));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>
//
// 
//
//
class IndexError : public AipsError {
public:
  //
  // Creates an GeneralIndexError and initializes the error message from
  // the parameter
  // <group>
  IndexError(const Char *str=0) : AipsError(str) {}
  IndexError(const String &str) : AipsError(str) {}
  // </group>

  //
  // This constructor is used by the exception mecanism to perform
  // type checking.
  //
  IndexError(ExcpError *excp);

  //
  // Destructor which does nothing.
  //
  ~IndexError();
};


// <summary>Index errors returning the bad index</summary>
// <use visibility=export>
//
// <reviewed reviewer="" date="" tests="" demos="">
// </reviewed>
//
// <synopsis>
// This class is templated to allow generalalized indexes to be returned
// with the error message i.e. the class is templated on the index type.
//
//  <note role=tip> The 
//         <linkto class=FreezeCleanup><src>FreezeCleanup</src> </linkto> 
//         and <linkto class=UnfreezeCleanup><src>UnfreezeCleanup</src>
//         </linkto> members are used to freeze and 
//         unfreeze the <linkto class=Cleanup><src>Cleanup</src></linkto>
//         mechanism. This is necessary because the templated key may or
//         may not be derived from <src>Cleanup</src>, and thus could be
//         destructed by the exception mechanism before the exception is
//         caught.
//
// </synopsis>
//
// <example>
// <srcblock>
//     throw(indexError<int>(3,"ANY STRING"));/
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>
//
// 
//
template<class t> class indexError : public IndexError {
protected:
  FreezeCleanup f;
  t oIndex;                 // Offending Index
  UnfreezeCleanup u;
public:
  //
  // This constructor takes the error message and the index
  // which cause the error to occur.
  //
  // <group>
  indexError(t oI, const Char *str=0);
  indexError(t oI, const String &str);
  // </group>

  //
  // This constructor is used by the exception mecanism to perform
  // type checking.
  //
  indexError(ExcpError *excp);

  //
  // Destructor which does nothing.
  //
  ~indexError();

};



// <summary>Duplicate key errors</summary>
// <use visibility=export>
//
// <reviewed reviewer="" date="" tests="" demos="">
// </reviewed>
//
// <synopsis>
// This class is the base class of all duplicate key errors. It is
// defined to allow the user to catch any of the many kinds of DuplErrors
// which may be thrown. It can also be thrown itself if returning
// the illegal key is unimportant.
// </synopsis>
//
// <example>
// <srcblock>
//    throw(DuplError("ANY STRING"));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>
//
class DuplError : public AipsError {
public:
  //
  // Creates an DuplError and initializes the error message from
  // the parameter
  // <group>
  DuplError(const Char *str=0) : AipsError(str) {}
  DuplError(const String &str) : AipsError(str) {}
  // </group>

  //
  // This constructor is used by the exception mecanism to perform
  // type checking.
  //
  DuplError(ExcpError *excp);

  //
  // Destructor which does nothing.
  //
  ~DuplError();

};

// <summary>Duplicate key errors where the bad key is returned</summary>
// <use visibility=export>
//
// <reviewed reviewer="" date="" tests="" demos="">
// </reviewed>
//
// <synopsis>
//  This template is for generalized duplicate key errors where the template
//  type parameter is the type of the key which caused the error. Because this
//  class is derived from <linkto class=DuplError><src>DuplError</src>
//  </linkto>, the user to catch all duplicate key errors with one catch
//  statement. 
//
//  <note role=tip> The 
//         <linkto class=FreezeCleanup><src>FreezeCleanup</src> </linkto> 
//         and <linkto class=UnfreezeCleanup><src>UnfreezeCleanup</src>
//         </linkto> members are used to freeze and 
//         unfreeze the <linkto class=Cleanup><src>Cleanup</src></linkto>
//         mechanism. This is necessary because the templated key may or
//         may not be derived from <src>Cleanup</src>, and thus could be
//         destructed by the exception mechanism before the exception is
//         caught.
//
// </synopsis>
//
// <example>
//     throw(duplError<int>(4,"ANY STRING"));
// </example>
//
// <todo asof="">
// </todo>
//
template<class t> class duplError : public DuplError {
protected:
  FreezeCleanup f;
  t oKey;                   // Offending Key
  UnfreezeCleanup u;
public:
  //
  // This constructs a "duplError" for the offending key, and an
  // optional character string.
  //
  // <group>
  duplError(t oI, const Char *str=0);
  duplError(t oI, const String &str);
  // </group>

  //
  // This constructor is used by the exception mecanism to perform
  // type checking.
  //
  duplError(ExcpError *excp);

  //
  // Destructor which does nothing.
  //
  ~duplError();

};



// <summary>Exception which halts execution</summary>
// <use visibility=export>
//
// <reviewed reviewer="" date="" tests="" demos="">
// </reviewed>
//
// <synopsis>
// This error causes an execution to halt regardless. It
// causes execution to halt before the exception can be caught.
// </synopsis>
//
// <example>
// <srcblock>
//     throw(AbortError("ANY STRING"));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>
//
class AbortError : public AipsError {
public:
  //
  // This constructs a "AbortError" from the error message.
  //
  // <group>
  AbortError(const Char *str);
  AbortError(const String &str);
  // </group>

  //
  // Destructor which does nothing.
  //
  ~AbortError();
};

// To make these error files known to the system..
#include <aips/Utilities/RegexError.h>

#endif
