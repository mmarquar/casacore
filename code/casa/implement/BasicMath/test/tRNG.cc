#include <aips/aips.h>
#include <aips/Exceptions/Error.h>
#include <aips/Utilities/String.h>
#include <iostream.h>
#include <iomanip.h>

#include <aips/Mathematics/Random.h>

int main() {
  try {
    uInt i;
    Float f;
    Double d;
    cout << "testing the MLCG generator" << endl;
    {
      cout << "random integers, floats & doubles" << endl; 
      MLCG g;
      for (uInt k = 0; k < 4; k++) {
	// Note the values are calculated here rather than in the print
	// statemement because of the problem discussed in
	// http://aips2.nrao.edu/mail/aips2-lib/1391
 	i = g.asuInt(); f = g.asFloat(); d = g.asDouble();
	cout << k << ": " << setbase(16) << i
 	     << ": " << setprecision(6) << f
 	     << ": " << setprecision(12) << d << endl;
      }
      cout << "resetting the generator. Should get the same numbers" << endl; 
      g.reset();
      for (uInt k = 0; k < 4; k++) {
 	i = g.asuInt(); f = g.asFloat(); d = g.asDouble();
 	cout << k << ": " << setbase(16) << i
 	     << ": " << setprecision(6) << f
 	     << ": " << setprecision(12) << d << endl;
      }
    }
    {
      cout << "Using a different seed" << endl; 
      MLCG g(1, 0);
      for (uInt k = 0; k < 4; k++) {
	i = g.asuInt(); f = g.asFloat(); d = g.asDouble();
	cout << k << ": " << setbase(16) << i 
	     << ": " << setprecision(6) << f 
	     << ": " << setprecision(12) << d << endl;
      }
      cout << "resetting the generator. Should get the same numbers" << endl; 
      g.reset();
      for (uInt k = 0; k < 4; k++) {
	i = g.asuInt(); f = g.asFloat(); d = g.asDouble();
	cout << k << ": " << setbase(16) << i 
	     << ": " << setprecision(6) << f 
	     << ": " << setprecision(12) << d << endl;
      }
    }
    cout << "testing the ACG generator" << endl;
    {
      cout << "random integers, floats & doubles" << endl; 
      ACG g;
      for (uInt k = 0; k < 4; k++) {
	i = g.asuInt(); f = g.asFloat(); d = g.asDouble();
	cout << k << ": " << setbase(16) << i 
	     << ": " << setprecision(6) << f 
	     << ": " << setprecision(12) << d << endl;
      }
      cout << "resetting the generator. Should get the same numbers" << endl; 
      g.reset();
      for (uInt k = 0; k < 4; k++) {
	i = g.asuInt(); f = g.asFloat(); d = g.asDouble();
	cout << k << ": " << setbase(16) << i 
	     << ": " << setprecision(6) << f 
	     << ": " << setprecision(12) << d << endl;
      }
    }
    {
      cout << "Using a different seed" << endl; 
      ACG g(7326458, 98);
      for (uInt k = 0; k < 4; k++) {
	i = g.asuInt(); f = g.asFloat(); d = g.asDouble();
	cout << k << ": " << setbase(16) << i 
	     << ": " << setprecision(6) << f 
	     << ": " << setprecision(12) << d << endl;
      }
      cout << "resetting the generator. Should get the same numbers" << endl; 
      g.reset();
      for (uInt k = 0; k < 4; k++) {
	i = g.asuInt(); f = g.asFloat(); d = g.asDouble();
	cout << k << ": " << setbase(16) << i 
	     << ": " << setprecision(6) << f 
	     << ": " << setprecision(12) << d << endl;
      }
    }
  }
  catch (AipsError x) {
    cerr << x.getMesg() << endl;
    cout << "FAIL" << endl;
    return 1;
  }
  catch (...) {
    cerr << "Exception not derived from AipsError" << endl;
    cout << "FAIL" << endl;
    return 1;
  }
  cout << "OK" << endl;
  return 0;
}
// Local Variables: 
// compile-command: "gmake tRNG"
// End: 
