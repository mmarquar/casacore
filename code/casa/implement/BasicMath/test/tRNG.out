testing the MLCG generator
random integers, floats & doubles
0: 4ffb0fe8: 0.951057: 0.568199365685
1: 7d0a8f71: 0.990539: 0.478558829083
2: 615c0703: 0.988474: 0.685491565147
3: 1d66a4ae: 0.0601412: 0.911177717968
resetting the generator. Should get the same numbers
0: 4ffb0fe8: 0.951057: 0.568199365685
1: 7d0a8f71: 0.990539: 0.478558829083
2: 615c0703: 0.988474: 0.685491565147
3: 1d66a4ae: 0.0601412: 0.911177717968
Using a different seed
0: 3e84478b: 0.916736: 0.219476846598
1: 3d4139c3: 0.114195: 0.0778288743032
2: 6636898c: 0.175281: 0.637837524635
3: 319ee548: 0.559797: 0.31682535589
resetting the generator. Should get the same numbers
0: 3e84478b: 0.916736: 0.219476846598
1: 3d4139c3: 0.114195: 0.0778288743032
2: 6636898c: 0.175281: 0.637837524635
3: 319ee548: 0.559797: 0.31682535589
testing the ACG generator
random integers, floats & doubles
0: 5ae6c126: 0.941752: 0.369242470763
1: f4cc99d2: 0.956651: 0.634494324782
2: 5c3b00ed: 0.418954: 0.929449494014
3: 627137fe: 0.913327: 0.00431547887706
resetting the generator. Should get the same numbers
0: 5ae6c126: 0.941752: 0.369242470763
1: f4cc99d2: 0.956651: 0.634494324782
2: 5c3b00ed: 0.418954: 0.929449494014
3: 627137fe: 0.913327: 0.00431547887706
Using a different seed
0: cbda4069: 0.913194: 0.472678594691
1: ea25a3d6: 0.915639: 0.0382038284907
2: 60bd8523: 0.0316063: 0.784168412303
3: 16a769f9: 0.0873321: 0.834734757385
resetting the generator. Should get the same numbers
0: cbda4069: 0.913194: 0.472678594691
1: ea25a3d6: 0.915639: 0.0382038284907
2: 60bd8523: 0.0316063: 0.784168412303
3: 16a769f9: 0.0873321: 0.834734757385
OK
